Git Remote Helper: E2E test harness

kim opened the pull request.

adinapoli-mndc commented on git-remote-ipfs/src/Network/IPFS/Git/RemoteHelper/Options.hs:
```diff
@@ -48,15 +41,9 @@ data IpfsPath
-       <> value (BaseUrl Http "localhost" 5001 mempty)
-       <> showDefault
-        )
+parseOptions = liftA2 Options
```
Interesting choice -- what made you prefer `liftA2` here?

adinapoli-mndc replied:
hah, fair enough 😁

adinapoli-mndc commented on git-remote-ipfs/test/e2e/Main.hs:
```diff
@@ -0,0 +1,79 @@
+            runGit_ ["push", "--quiet", "ipfs", "master"]
+            Text.stripEnd <$> runGit  ["remote", "get-url", "ipfs"]
+
+    withRepository' cliFactory (ropts (clonePath </> ".git") False) $ do
```
Perhaps is outside the scope of these end-2-end tests, but is there any `Assertion` we could meaningfully perform at this stage, after we exist the second `withRepository'` block?

kim replied:
See below - the assertion is `git diff -q` exiting `0`

kim commented on git-remote-ipfs/test/e2e/Main.hs:
```diff
@@ -0,0 +1,79 @@
+        runGit_ ["clone", "--quiet", url, Text.pack clonePath]
+        runGit_ ["remote", "add", "src", Text.pack repoPath]
+        runGit_ ["fetch", "--quiet", "src"]
+        runGit_ ["diff", "--quiet", "origin/master", "src/master"]
```
`--quiet` implies `--exit-code`, just like the regular `diff` command. `runGit_` throws an exception if the exit code /= 0, so the test fails.

Arguably, this whole `gitlib` stuff doesn’t pay for itself - the dependency footprint is pretty hefty - so I’m planning to move to using plain `typed-process` in the next PR. I’ll need it anyway to run `ipfs` commands. `assertBool “Source and clone differ” exitCode` presumably reads nicer.

geigerzaehler replied:
Thanks for clarifying this. Maybe a comment that the diff command is actually an assertion might be helpful.

kim commented:
@adinapoli-mndc @geigerzaehler PTAL

I switched to plain `typed-process` instead of the somewhat useless `gitlib-cmdline`, and added a smoke test utilising IPNS. 

Couldn't find a way around `gitlib-libgit2` for building the git history that would be less annoying than depending on OpenSSL and `text-icu`.

adinapoli-mndc commented:
👍 LGTM -- indeed it's a bit of a shame to depend on `text-icu` and `OpenSSL` , especially as the second tends not to be installed by default of most user desktops, but I guess that's the current state of affairs and we should roll with the punches in lack of a better alternative.

geigerzaehler commented:
> Is the `ipfs-test-network` image publicly docker-pullable?

It soon will be. I’ll make the `gcr.io/opensourcecoin` registry public and upload the image there.
